src/tensors/diagtensor.jl

```diff
@@ -0,0 +1,279 @@
+                               f₂::FusionTree{I,1}) where {I<:Sector}
+    s = f₁.uncoupled[1]
+    s == f₁.coupled == f₂.uncoupled[1] == f₂.coupled || throw(SectorMismatch())
+    return block(d, s)
+    # TODO: do we want a StridedView here? Then we need to allocate a new matrix.
+end
+
+function Base.setindex!(d::DiagonalTensorMap,
+                        v,
+                        f₁::FusionTree{I,1},
+                        f₂::FusionTree{I,1}) where {I<:Sector}
+    return copy!(getindex(d, f₁, f₂), v)
+end
+
+function Base.getindex(d::DiagonalTensorMap)
+    sectortype(d) === Trivial || throw(SectorMismatch())
+    return Diagonal(d.data)
+end
+
+# Index manipulations
+# -------------------
+function has_shared_permute(d::DiagonalTensorMap, (p₁, p₂)::Index2Tuple)
+    if p₁ === (1,) && p₂ === (2,)
+        return true
+    elseif p₁ === (2,) && p₂ === (1,) # transpose
+        return sectortype(d) === Trivial
+    else
+        return false
+    end
+end
+
+function permute(d::DiagonalTensorMap, (p₁, p₂)::Index2Tuple{1,1};
+                 copy::Bool=false)
+    if p₁ === (1,) && p₂ === (2,)
+        return copy ? Base.copy(d) : d
+    elseif p₁ === (2,) && p₂ === (1,) # transpose
+        if has_shared_permute(d, (p₁, p₂)) # tranpose for bosonic sectors
+            return DiagonalTensorMap(copy ? Base.copy(d.data) : d.data, dual(d.domain))
+        end
+        d′ = typeof(d)(undef, dual(d.domain))
+        for (c, b) in blocks(d)
+            f = only(fusiontrees(codomain(d), c))
+            ((f′, _), coeff) = only(permute(f, f, p₁, p₂))
+            c′ = f′.coupled
+            scale!(block(d′, c′), b, coeff)
+        end
+        return d′
+    else
+        throw(ArgumentError("invalid permutation $((p₁, p₂)) for tensor in space $(space(d))"))
+    end
+end
+
+# VectorInterface
+# ---------------
+function VectorInterface.zerovector(d::DiagonalTensorMap, ::Type{S}) where {S<:Number}
+    return DiagonalTensorMap(zerovector(d.data, S), d.domain)
+end
+function VectorInterface.add(ty::DiagonalTensorMap, tx::DiagonalTensorMap,
+                             α::Number, β::Number)
+    domain(ty) == domain(tx) || throw(SpaceMismatch("$(space(ty)) ≠ $(space(tx))"))
+    T = VectorInterface.promote_add(ty, tx, α, β)
+    return add!(scale!(zerovector(ty, T), ty, β), tx, α) # zerovector instead of similar preserves diagonal structure
+end
+
+# Linear Algebra and factorizations
+# ---------------------------------
+function one!(d::DiagonalTensorMap)
+    fill!(d.data, one(eltype(d.data)))
+    return d
+end
+function Base.one(d::DiagonalTensorMap)
+    return DiagonalTensorMap(one.(d.data), d.domain)
+end
+function Base.zero(d::DiagonalTensorMap)
+    return DiagonalTensorMap(zero.(d.data), d.domain)
+end
+
+function compose(d1::DiagonalTensorMap, d2::DiagonalTensorMap)
+    d1.domain == d2.domain || throw(SpaceMismatch())
+    return DiagonalTensorMap(d1.data .* d2.data, d1.domain)
+end
```

Comment:
Should this implementation be moved to `mul!(::DiagonalTensorMap, ::DiagonalTensorMap, ::DiagonalTensorMap)`, and use `compose_dest` to figure out that the result is `Diagonal`?

Reply:
We can also specialise `mul!(::DiagonalTensorMap, ::DiagonalTensorMap, ::DiagonalTensorMap)`, although already now this will lead to `mul!(::Diagonal,::Diagonal,::Diagonal)` on the blocks, so it is pretty good. But indeed, it can be faster since the spaces need to be the same and so it can just act on the vector as a whole without iterating over the blocks.

Reply:
Mostly, I meant that hoisting the implementation to `mul!` is slightly more generic, since that would then presumably also be used in the `tensorcontract!` functions, which use `mul!` instead of `compose`?